webcam-capture-drivers/driver-ffmpeg-cli/src/main/java/com/github/sarxos/webcam/ds/ffmpegcli/FFmpegCliDevice.java

```diff
@@ -1,260 +1,218 @@
+
+        Dimension resolution = getResolution();
+        image = new BufferedImage(resolution.width, resolution.height, BufferedImage.TYPE_3BYTE_BGR);
+    }
+
+    public void startProcess() throws IOException {
+        ProcessBuilder builder = new ProcessBuilder(buildCommand());
+        builder.redirectErrorStream(true); // so we can ignore the error stream
+
+        Process process = builder.start();
+        InputStream out = process.getInputStream();
+        OutputStream in = process.getOutputStream();
+
+        int SIZE = resolution.width * resolution.height * 3;
+        int CHUNK_SIZE = SIZE / 20;
+
+        int cursor = 0;
+
+        byte[] chunk = new byte[CHUNK_SIZE];
+        ByteBuffer buffer = ByteBuffer.allocate(SIZE);
+
+        while (isAlive(process)) {
+            int no = out.available();
+            if (no >= chunk.length) {
+                int n = out.read(chunk, 0, chunk.length);
+                cursor += n;
+
+                // If buffer is not full yet
+                if (cursor < SIZE) {
+                    // Append the chunk to the buffer
+                    buffer.put(chunk, 0, n);
+                } else {
+                    // Read Image
+                    byte[] bgr = buffer.array();
+                    image.setData(Raster.createRaster(image.getSampleModel(), new DataBufferByte(bgr, bgr.length), new Point()));
+
+                    buffer.clear();
+                    cursor = 0;
+                }
+            }
+        }
+    }
+
+    public static boolean isAlive(Process p) {
+        try {
+            p.exitValue();
+            return false;
+        } catch (IllegalThreadStateException e) {
+            return true;
+        }
+    }
+
+    private Dimension[] readResolutions(String res) {
+        List<Dimension> resolutions = new ArrayList<Dimension>();
+        String[] parts = res.split(" ");
+
+        for (String part : parts) {
+            String[] xy = part.split("x");
+            resolutions.add(new Dimension(Integer.parseInt(xy[0]), Integer.parseInt(xy[1])));
+        }
+
+        return resolutions.toArray(new Dimension[resolutions.size()]);
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public Dimension[] getResolutions() {
+        return resolutions;
+    }
+
+    @Override
+    public Dimension getResolution() {
+        if (resolution == null) {
+            resolution = getResolutions()[0];
+        }
+        return resolution;
+    }
+
+    private String getResolutionString() {
+        Dimension d = getResolution();
+        return String.format("%dx%d", d.width, d.height);
+    }
+
+    @Override
+    public void setResolution(Dimension resolution) {
+        this.resolution = resolution;
+    }
+
+    public void open() {
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                open.set(true);
```

Comment:
This is very susceptible for race conditions. This about this scenario:

1. The thread here run and set open = true,
2. The `startProcess()` takes some time to take over and creation of `Process` is delayed,
3. In the meantime someone calls `close()`,
4. Result - NPE in `close()` method caused by process field being `null`.

To avoid this you should introduce some kind of lock which block for the time when device is initialized and before it's _really_ open, for example (not tested, wrote from memory):

```java

final CountDownLatch latch = new CountDownLatch(1);

Thread t = new Thread(new Runnable() {
  public void run() {
    startProcess(latch);
  }
});
t.setDaemon(true);
t.start();

latch.await();
```

And in the `startProcess()`:

```java
public void startProcess(final CountDownLatch latch) throws IOException {
  ...
  image.setData(...)
  latch.countDown(); // no matter how many times it's invoked, after it reaches zero
}
```